DDRec/src/plugins/addVariantParameters.cpp

```diff
@@ -0,0 +1,120 @@
+//--------------------------------------------------------------------------
+// Copyright (C) Organisation europeenne pour la Recherche nucleaire (CERN)
+// All rights reserved.
+//
+// For the licensing terms see $DD4hepINSTALL/LICENSE.
+// For the list of contributors see $DD4hepINSTALL/doc/CREDITS.
+//
+// Author     : P.Gessinger
+//
+//==========================================================================
+#include "DD4hep/Detector.h"
+#include "DD4hep/Factories.h"
+#include "DD4hep/Printout.h"
+#include "DD4hep/DetectorTools.h"
+
+#include "DDRec/DetectorData.h"
+
+#include <regex>
+
+
+namespace dd4hep::rec {
+
+/**
+ *  @author P.Gessinger, CERN
+ *  @date May, 25 2022
+ *
+ *  \brief Plugin which attaches parameters with string keys and int, double, 
+ *  string or bool values to detector elements
+ *
+ *  This plugin can be configured to attach a varying number of variant paremeters
+ *  to a single detector elements. The first argument is the detector element name.
+ *  Any following arguments should conform to the following format:
+ *    
+ *    key: type = value
+ *  
+ *  where type can be one of "int", "bool", "double", or "str", with "str" being the default.
+ *  This will ensure an instance of the @c VariantParameters extension is attached to a detector
+ *  element, and insert @c key with a @c value of type @c type.
+ */
+
+static long addVariantParameters(Detector& description, int argc, char** argv) {
+
+  if(argc < 2) {
+    printout(ERROR,"ParametersPlugin","+++ Invalid arguments");
+    return 0;
+  }
+
+  printout(INFO,"ParametersPlugin","+++ Applying %d parameters", argc);
+
+
+  static std::regex expr{"^(\\w+)(?:: ?(.*?)?)? ?= ?(.*)"};
+
+  std::string detector = argv[0];
+
+  for(int i=1;i<argc;i++) {
+    std::string kv = argv[i];
+    std::smatch m;
+    if(!std::regex_match(kv, m, expr)) {
+      throw std::runtime_error("Unable to parse key-value pair to assign");
+    }
+    if (m.size() != 4) {
+      throw std::runtime_error("Unable to parse key-value pair to assign");
+    }
+
+    std::string key = m.str(1);
+    std::string type = m.str(2);
+    if (type == "") {
+      type = "str";
+    }
+    std::string value = m.str(3);
+
+    printout(INFO,"ParametersPlugin","+++ %s -> %s: %s = %s", detector.c_str(), key.c_str(), type.c_str(), value.c_str());
+
+    DetElement elt = description.detector(detector);
+    if(!elt.isValid()){
+      printout(ERROR,"ParametersPlugin","+++ Did not find element with name '%s'", detector.c_str());
+      return 0;
+    }
+
+    VariantParameters* extension = elt.extension<VariantParameters>(false);
+    if(extension == nullptr) {
+      extension = new VariantParameters();
+      elt.addExtension<VariantParameters>(extension);
+    }
+
+    if (type == "str" || type == "string") {
+      extension->variantParameters[key] = value;
+    }
+    else if (type == "double") {
+      extension->variantParameters[key] = std::stod(value);
```

Comment:
I think if we switch to the dd4hep parsing functions, we can use things like units, constants, whatever is known to the grammar, in the values? Or is that already working?

```suggestion
      extension->variantParameters[key] = dd4hep:_toDouble(value);
```
https://dd4hep.web.cern.ch/dd4hep/reference/group__DD4HEP__CORE.html#gab6fd7f4fe0ac9fa2c4f41132be909103

Reply:
Ah I didn't consider using the full parsing functions. I can look into that though. Does the evaluator support anything other than double results, though? Reading the code it seems to me like it's restricted to double, but here we'd probably at least support string and int values.

Reply:
yes, keeping the `type: value`, but going through the dd4hep parsing for ints and doubles.
Not sure it would make sense for strings

Reply:
Ah right I see, I didn't realize `_toDouble` and `_toInt` already run the evaluator.

This should be good then, right?